### PR TITLE
Issue #31: docker flags

### PR DESCRIPTION
**Pull request by gfechio**
Added new command line for connection:
- `--target-ssh`
- `--target-container`
- `--target-localhost`
Sub-commands `sudo` and `disable-agent` now under `--target-options` separated by commas.
Closes issue #31

### CODE REVIEW

File: cmd/lib.go [hunk 1]
```diff
@@ -30,57 +30,65 @@ var defaultDisableAgent = false
...
-		&sudo, "sudo", "", defaultSudo,
-		"Use sudo when interacting with host",
-	)
-
-	cmd.Flags().BoolVarP(
-		&disableAgent, "disable-agent", "", defaultDisableAgent,
-		"Disables copying temporary a small agent to remote hosts. This can make things very slow, as without the agent, iteraction require running multiple commands. The only (unusual) use case for this is when the host architecture is not supported by the agent.",
+		&options, "target-options", "o", defaultDocker,
+		"Usage: --target-options=sudo,disable-agent \n"+
+			"Applies options to your connection: \n"+
```
Comment (gfechio) [inline cmd/lib.go:80]: Review this as I am not certain it is the most comprehensive approach.

File: cmd/lib_linux.go [hunk 2]
```diff
@@ -24,21 +24,21 @@ func GetHost(ctx context.Context) (host.Host, error) {
...
+		return nil, errors.New("no target host specified: must pass either --target-localhost, --target-ssh or --target-docker")
 	}
 
 	return wrapHost(ctx, hst)
 }
 
 func AddHostFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVarP(
-		&localhost, "localhost", "", defaultLocalhost,
+		&localhost, "target-localhost", "1", defaultLocalhost,
```
Comment (gfechio) [inline cmd/lib_linux.go:41]: This short-hand seems quite odd IMO, but I am not sure what would be the best approach here, maybe leave it without a short-hand?

File: internal/host/docker_unix.go [hunk 3]
```diff
@@ -84,7 +85,11 @@ func (d Docker) Run(ctx context.Context, cmd host.Cmd) (host.WaitStatus, error)
 }
 
 func (d Docker) String() string {
-	return fmt.Sprintf("%s@docker:%s", d.User, d.Container)
+	return fmt.Sprintf(d.Connection)
+}
+
+func (d Docker) Type() string {
+	return "docker"
```
Comment (gfechio) [inline internal/host/docker_unix.go:92]: Fix for Host.Type()

File: cmd/lib_linux.go (hunk 2 above)
Comment (fornellas) [reply cmd/lib_linux.go:41]: TBH, IDK. These can me

File: host/host.go [hunk 4]
```diff
@@ -61,6 +61,9 @@ type Host interface {
 	// A string representation of the host which uniquely identifies it, eg, its FQDN.
 	String() string
 
+	// String representation for the type of connection is used. eg: ssh, localhost, docker
+	Type() string
```
Comment (fornellas) [inline host/host.go:65]: I wonder if we'd be better off using good names for the `Host` types, and then:

```go
func GetHostType(hst Host) string {
  return strings.ToLower(reflect.TypeOf(hst).Elem().Type().Name)
}
```

Comment (fornellas) [reply host/host.go:65]: Oh, this may be an issue with `h.Host.Type()` below...

File: internal/host/docker_unix.go [hunk 5]
```diff
@@ -15,16 +15,13 @@ import (
 // Docker uses docker exec to target a running container.
 type Docker struct {
 	cmdHost
-	// Container name
-	Container string
-	// Username or UID (format: "<name|uid>[:<group|gid>]")
-	User string
+	// Username or UID (format: "<name|uid>[:<group|gid>]") and Container Name ( eg: root@ubuntu )
```
Comment (fornellas) [inline internal/host/docker_unix.go:18]: WDYT?

```suggestion
	// User/group and image in the format "<name|uid>[:<group|gid>@<image>" (eg: root@ubuntu)
```

File: internal/host/docker_unix.go [hunk 6]
```diff
@@ -15,16 +15,13 @@ import (
...
+	Connection string
```
Comment (fornellas) [inline internal/host/docker_unix.go:19]: Nit: This name sounded odd to me at first... `docker exec` has `--user` (which oddly can be... a group as well...)) and `IMAGE`, and... there's not really a connection going on (necessarily). I don't have a better name though...

File: internal/host/docker_unix.go [hunk 7]
```diff
@@ -55,9 +52,13 @@ func (d Docker) Run(ctx context.Context, cmd host.Cmd) (host.WaitStatus, error)
 	if cmd.Stdin != nil {
 		args = append(args, "--interactive")
 	}
-	args = append(args, []string{"--user", d.User}...)
+
+	dockerConnectionUser := strings.Split(d.Connection, "@")[0]
```
Comment (fornellas) [inline internal/host/docker_unix.go:56]: this and the other one can be simpler: this is within docker already... and is not a connection anymore.
```suggestion
	user := strings.Split(d.Connection, "@")[0]
```

File: internal/host/docker_unix.go [hunk 8]
```diff
@@ -84,7 +85,11 @@ func (d Docker) Run(ctx context.Context, cmd host.Cmd) (host.WaitStatus, error)
 }
 
 func (d Docker) String() string {
-	return fmt.Sprintf("%s@docker:%s", d.User, d.Container)
```
Comment (fornellas) [inline internal/host/docker_unix.go:87]: this used to be TERRIBLE, lol

File: go.mod [hunk 9]
```diff
@@ -1,6 +1,6 @@
 module github.com/fornellas/resonance
 
-go 1.22.5
+go 1.22.6
```
Comment (fornellas) [inline go.mod:3]: intentional or not... harmless :-D go go go!

File: cmd/validate.go [hunk 10]
```diff
@@ -27,7 +27,7 @@ var ValidateCmd = &cobra.Command{
 		}
 		defer hst.Close()
 
-		logger.Info("🔍 Validating", "path", path, "host", hst)
+		logger.Info("🔍 Validating", "path", path, hst.Type(), hst)
```
Comment (fornellas) [inline cmd/validate.go:30]: ah...`slog`... So, the interface is... a bit odd, I reckon: `Info(message, key1, value1, key2, value2 etc)`.

so, this would make more sense I think:

```suggestion
		logger.Info("🔍 Validating", "path", path, "target", hst)
```

so, it'll use `Host.String()`. BUT, thiss would hide the type, so it'd either have to be something like:

- `fmt.Sprintf("%s:%s", hst.Type(), hst.Strin())`
- Update all `Host.String()` to do that.

WDYT?

File: cmd/lib_linux.go (hunk 2 above)
Comment (fornellas) [reply cmd/lib_linux.go:45]: I had to unresolve, sorry, I'm sure you've seen this, lol: https://github.com/ltworf/funny-manpages/blob/master/man/sex.6fun#L16.

File: cmd/lib.go [hunk 11]
```diff
@@ -104,10 +108,8 @@ func getStoreCommon(hst host.Host) storePkg.Store {
 func init() {
 	resetFlagsFns = append(resetFlagsFns, func() {
 		ssh = defaultSsh
-		dockerContainer = defaultDockerContainer
-		dockerUser = defaultDockerUser
-		sudo = defaultSudo
-		disableAgent = defaultDisableAgent
+		docker = defaultDocker
+		options = defaultOptions
```
Comment (fornellas) [inline cmd/lib.go:112]: worth a word on why this fuc@!$!@ is here: Cobra is great, but when testing, it does NOT reset these... and one test can taint the results of another... in the absence of a better solution, this is what we have.

File: cmd/lib.go [hunk 12]
```diff
@@ -17,17 +17,11 @@ var Exit func(int) = func(code int) { os.Exit(code) }
 var ssh string
 var defaultSsh = ""
 
-var dockerContainer string
-var defaultDockerContainer = ""
+var docker string
```
Comment (fornellas) [inline cmd/lib.go:20]: nit: do you mind changing all of these to match the long format option it refers to, for consistency?

```suggestion
var targetDocker string
```

File: internal/host/docker_unix.go (hunk 5 above)
Comment (fornellas) [reply internal/host/docker_unix.go:18]: PS: see comment below as well, this may need updating.

File: internal/host/docker_unix.go [hunk 13]
```diff
@@ -55,9 +52,13 @@ func (d Docker) Run(ctx context.Context, cmd host.Cmd) (host.WaitStatus, error)
 	if cmd.Stdin != nil {
 		args = append(args, "--interactive")
 	}
-	args = append(args, []string{"--user", d.User}...)
+
+	dockerConnectionUser := strings.Split(d.Connection, "@")[0]
+	dockerConnectionContainer := strings.Split(d.Connection, "@")[1]
```
Comment (fornellas) [inline internal/host/docker_unix.go:57]: This can panic: if a user gives invalid input (eg: `container_name_only`), the length of `strings.Split(d.Connection, "@")` will be 1, and this will panic.

Originally, we had user/group default to `0:0` (root). I think we could keep this, as most cases, we'd be running as root. So, the format would be: `[<name|uid>[:<group|gid>@]<image>`.

I'm also thinking whether it is best to move the logic to parse this separately, whether than here on `NewDocker`...

File: host/host.go (hunk 4 above)
Comment (gfechio) [reply host/host.go:65]: That seems like a more complex approach, and the use of `reflect` might be a overkill, at least for now.
We can definitely implement it, but I'd like to mature this idea a bit more.

File: cmd/lib_linux.go (hunk 2 above)
Comment (gfechio) [reply cmd/lib_linux.go:45]: It is indeed a funny page, but in a serious note you can find similar approaches in linux basic commands. 
https://man7.org/linux/man-pages/man1/ls.1.html

File: internal/host/docker_unix.go (hunk 6 above)
Comment (gfechio) [reply internal/host/docker_unix.go:19]: I was thinking about `ConnectionString` such as used for Postgres, MySQL and some times Redis. That would make it bit more comprehensive. WDYT ?

File: cmd/lib.go [hunk 14]
```diff
@@ -36,48 +30,58 @@ var defaultStoreHostTargetPath = "/var/lib/resonance"
 
 func wrapHost(ctx context.Context, hst host.Host) (host.Host, error) {
 	var err error
-	if sudo {
+	switch options {
```
Comment (fornellas) [inline cmd/lib.go:33]: I think this logic changes some of the rules, which may have been unintentional.

Before:

- sudo on request
- agent default:
	- localhost: NO (because localhost is faster without an agent)
	- docker: NO (it should be faster with agent, this should be changed to yes)
	- ssh: YES (as it is a billion times faster with agent)

basically, the idea is "do what's fasted by default" (including the docker agent enabled fix :-D), and only disable agent for odd cases.

How about something like this?

```go
optionsMap := map[string]bool{
	"sudo": false,
	"disable-agent": false,
}
for _, o := range strings.Split(options, ",") {
	if _, ok := optionsMap[o] ; !ok {
		return fmt.Errorf("invalid option: %s", o)
	}
	optionsMap[o] = true
}

if optionsMap["sudo"] {
	// enable sudo
}

if hst.Type() != "localhost" && ! options["disable-agent"] {
	// enable agent
}
```

File: cmd/lib.go [hunk 15]
```diff
@@ -36,48 +30,58 @@ var defaultStoreHostTargetPath = "/var/lib/resonance"
...
+		&ssh, "target-ssh", "s", defaultSsh,
 		"Applies configuration to given hostname using SSH in the format: [<user>[;fingerprint=<host-key fingerprint>]@]<host>[:<port>]",
 	)
 
 	cmd.Flags().StringVarP(
-		&dockerContainer, "docker-container", "", defaultDockerContainer,
-		"Applies configuration to given Docker container name",
+		&docker, "target-docker", "d", defaultDocker,
+		"Applies configuration to given Docker container name \n"+
+			"Use given format 'USER@CONTAINER_ID'",
```
Comment (fornellas) [inline cmd/lib.go:75]: NO NEED TO YELL, lol, I think the string here should match what's at the Host implementantion (<user> bla bla bla).

File: cmd/lib.go [hunk 16]
```diff
@@ -36,48 +30,58 @@ var defaultStoreHostTargetPath = "/var/lib/resonance"
...
-		&ssh, "ssh", "", defaultSsh,
+		&ssh, "target-ssh", "s", defaultSsh,
 		"Applies configuration to given hostname using SSH in the format: [<user>[;fingerprint=<host-key fingerprint>]@]<host>[:<port>]",
 	)
 
 	cmd.Flags().StringVarP(
-		&dockerContainer, "docker-container", "", defaultDockerContainer,
-		"Applies configuration to given Docker container name",
+		&docker, "target-docker", "d", defaultDocker,
+		"Applies configuration to given Docker container name \n"+
```
Comment (fornellas) [inline cmd/lib.go:74]: I haven't tested Cobra with `\n`, does it properly format things, or it is odd looking?

File: cmd/lib.go [hunk 17]
```diff
@@ -36,48 +30,58 @@ var defaultStoreHostTargetPath = "/var/lib/resonance"
...
-	cmd.Flags().BoolVarP(
-		&sudo, "sudo", "", defaultSudo,
-		"Use sudo when interacting with host",
-	)
-
-	cmd.Flags().BoolVarP(
-		&disableAgent, "disable-agent", "", defaultDisableAgent,
-		"Disables copying temporary a small agent to remote hosts. This can make things very slow, as without the agent, iteraction require running multiple commands. The only (unusual) use case for this is when the host architecture is not supported by the agent.",
+		&options, "target-options", "o", defaultDocker,
+		"Usage: --target-options=sudo,disable-agent \n"+
```
Comment (fornellas) [inline cmd/lib.go:79]: I'd simplify this to something like: _comma separated list of target options: "sudo", to run as root via sudo; "disable-agent", disable ephemeral agent usage (MUCH slower)_

**Review by fornellas — CHANGES_REQUESTED** [review]
LPGTM. There are some nits (take with as many grains of salt as you'd like), some questions (open, I'm not opinionated), but I think the options parsing needs amending before merging. Feel _empowered_ to disagree / challenge :-D

File: cmd/lib_linux.go (hunk 2 above)
Comment (fornellas) [reply cmd/lib_linux.go:45]: Yes I know, I liked your solution. But `--target-localhost` sort feels like `sex -1`, lol.

File: internal/host/docker_unix.go (hunk 6 above)
Comment (fornellas) [reply internal/host/docker_unix.go:19]: After writing some of the comments below, regarding parsing this, I wonder if keeping user & image here as before, and just moving this logic to the frontend (CLI) would be better. Naming stuff is hard, lol, doing so avoids this conversation altogether.
(don't forget, it is a nit comment)

File: cmd/lib.go (hunk 12 above)
Comment (gfechio) [reply cmd/lib.go:20]: It this is changed, makes sense to keep consistency across variables and call 
-`targetSSH`
-`targetDocker`
-`targetLocalhost`

File: internal/host/docker_unix.go (hunk 6 above)
Comment (gfechio) [reply internal/host/docker_unix.go:19]: We need to remember, this is not the IMAGE name, it is the container name or ID. Better not mix taxonomies.
Now I am okay with `ConnectionString`, it makes much more sense anyway.

File: internal/host/docker_unix.go (hunk 7 above)
Comment (gfechio) [reply internal/host/docker_unix.go:56]: Personal preference here, but it is much easier to have self-explanatory variable names according to it's scope, as this user although a user in the general purpose it is strongly connected to the **docker** scope.

File: cmd/lib.go (hunk 11 above)
Comment (gfechio) [reply cmd/lib.go:112]: I don't quite follow what you mean here :stuck_out_tongue:

File: cmd/lib.go (hunk 14 above)
Comment (gfechio) [reply cmd/lib.go:33]: That looks like a more elegant solution indeed.

File: cmd/lib.go (hunk 16 above)
Comment (gfechio) [reply cmd/lib.go:74]: I did no worries :blinks:

File: cmd/lib.go (hunk 14 above)
Comment (gfechio) [reply cmd/lib.go:33]: :heavy_check_mark:

File: cmd/lib.go (hunk 15 above)
Comment (gfechio) [reply cmd/lib.go:75]: :heavy_check_mark:

File: internal/host/docker_unix.go (hunk 6 above)
Comment (fornellas) [reply internal/host/docker_unix.go:19]: I still don't have a better name, and... not worth the discussion: go for whatever you think is best, I trust you :-D

File: internal/host/docker_unix.go (hunk 7 above)
Comment (fornellas) [reply internal/host/docker_unix.go:56]: however you prefer :-D

File: cmd/lib.go (hunk 11 above)
Comment (fornellas) [reply cmd/lib.go:109]: I just wanted to explain why this code exist, it does look odd even to me who wrote it, lol.

File: cmd/lib.go (hunk 12 above)
Comment (fornellas) [reply cmd/lib.go:23]: remember, nit == take it or leave it, no problem.

File: cmd/lib.go [hunk 18]
```diff
@@ -36,18 +33,31 @@ var defaultStoreHostTargetPath = "/var/lib/resonance"
...
-		hst, err = ihost.NewSudo(ctx, hst)
-		if err != nil {
-			return nil, err
-		}
+
+	optionsMap := map[string]bool{
+		"sudo":          false,
+		"disable-agent": false,
 	}
+	if options != "" {
```
Comment (fornellas) [inline cmd/lib.go:41]: I think we need to drop this, because otherwise the default (options == ""), we won't go inside this if, and this won't run:

```
		if hst.Type() != "localhost" && !optionsMap["disable-agent"] {
			hst, err = ihost.NewAgent(ctx, hst)
			if err != nil {
				return nil, err
			}
		}
```

so we wouldn't be enabling the agent by default (on the necessary cases).

File: cmd/lib.go [hunk 19]
```diff
@@ -56,28 +66,20 @@ func wrapHost(ctx context.Context, hst host.Host) (host.Host, error) {
...
+		&ssh, "target-ssh", "s", defaultSsh,
 		"Applies configuration to given hostname using SSH in the format: [<user>[;fingerprint=<host-key fingerprint>]@]<host>[:<port>]",
 	)
 
 	cmd.Flags().StringVarP(
-		&dockerContainer, "docker-container", "", defaultDockerContainer,
-		"Applies configuration to given Docker container name",
+		&docker, "target-docker", "d", defaultDocker,
+		"Applies configuration to given Docker container name \n"+
+			"Use given format '[<user>]@<container_name>|<container_id>'",
```
Comment (fornellas) [inline cmd/lib.go:76]: I think this may need to match the format on `internal/host/docker_unix.go`:

```suggestion
			"Use given format '[<name|uid>[:<group|gid>]@]<image>'",
```

File: internal/host/docker_unix.go [hunk 20]
```diff
@@ -15,16 +15,13 @@ import (
 // Docker uses docker exec to target a running container.
 type Docker struct {
 	cmdHost
-	// Container name
-	Container string
-	// Username or UID (format: "<name|uid>[:<group|gid>]")
-	User string
+	// User/group and image in the format "<name|uid>[:<group|gid>@<image>" (eg: root@ubuntu)
```
Comment (fornellas) [inline internal/host/docker_unix.go:18]:
```suggestion
	// User/group and image in the format "[<name|uid>[:<group|gid>]@]<image>" (eg: root@ubuntu)
```

File: internal/host/docker_unix.go [hunk 21]
```diff
@@ -55,9 +52,19 @@ func (d Docker) Run(ctx context.Context, cmd host.Cmd) (host.WaitStatus, error)
...
-	args = append(args, []string{"--user", d.User}...)
+
+	parts := strings.Split(d.ConnectionString, "@")
+
+	if len(parts) < 2 {
+		return host.WaitStatus{}, fmt.Errorf("invalid connection string format")
+	}
+
+	dockerConnectionUser := parts[0]
+	dockerConnectionContainer := parts[1]
```
Comment (fornellas) [inline internal/host/docker_unix.go:63]: I think we may be missing the case where we just get the container (without `user@` prefix), and the error:

```suggestion
	var dockerConnectionUser, dockerConnectionContainer string
	switch len(parts) {
	case 1:
		dockerConnectionUser = "0:0"
		dockerConnectionContainer = parts[0]
	case 2:
		dockerConnectionUser = parts[0]
		dockerConnectionContainer = parts[1]
	default:
		return host.WaitStatus{}, fmt.Errorf("invalid connection string format: must match [<name|uid>[:<group|gid>]@<image>, got %s", d.ConnectionString)
	}
```

**Review by fornellas — CHANGES_REQUESTED** [review]
image

File: cmd/lib.go (hunk 18 above)
Comment (gfechio) [reply cmd/lib.go:41]: without it the binary won't run, it outputs: `Error invalid command`

Comment (gfechio) [reply cmd/lib.go:41]: image

Comment (fornellas) [reply cmd/lib.go:41]: this should do it:

```diff
diff --git a/cmd/lib.go b/cmd/lib.go
index a4b8e9e..dd04d2d 100644
--- a/cmd/lib.go
+++ b/cmd/lib.go
@@ -38,26 +38,27 @@ func wrapHost(ctx context.Context, hst host.Host) (host.Host, error) {
                "sudo":          false,
                "disable-agent": false,
        }
+
        if options != "" {
                for _, o := range strings.Split(options, ",") {
                        if _, ok := optionsMap[o]; !ok {
-                               return nil, fmt.Errorf("invalid option: %s", o)
+                               return nil, fmt.Errorf("invalid option: %#v", o)
                        }
                        optionsMap[o] = true
                }
+       }
 
-               if optionsMap["sudo"] {
-                       hst, err = ihost.NewSudo(ctx, hst)
-                       if err != nil {
-                               return nil, err
-                       }
+       if optionsMap["sudo"] {
+               hst, err = ihost.NewSudo(ctx, hst)
+               if err != nil {
+                       return nil, err
                }
+       }
 
-               if hst.Type() != "localhost" && !optionsMap["disable-agent"] {
-                       hst, err = ihost.NewAgent(ctx, hst)
-                       if err != nil {
-                               return nil, err
-                       }
+       if hst.Type() != "localhost" && !optionsMap["disable-agent"] {
+               hst, err = ihost.NewAgent(ctx, hst)
+               if err != nil {
+                       return nil, err
                }
        }
```

but then, it is reaching the stack limit, due to infinite recursive call. Separate issue, too tired to debug this today, but it kinda looks like there's a docker host wrapping itself somehow...

```
✔fornellas@brown:~/src/resonance ((661c73b...) *$%)$ ./resonance.linux.amd64 validate --target-docker foo@bar foo.yaml 
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0xc0201e0390 stack=[0xc0201e0000, 0xc0401e0000]
fatal error: stack overflow

runtime stack:
runtime.throw({0xb08d6c?, 0x200000001?})
        /home/fornellas/.cache/resonance/GOROOT/go1.22.6.linux-amd64/src/runtime/panic.go:1023 +0x5c fp=0x7fffe456eba8 sp=0x7fffe456eb78 pc=0x43f37c
runtime.newstack()
        /home/fornellas/.cache/resonance/GOROOT/go1.22.6.linux-amd64/src/runtime/stack.go:1103 +0x5bd fp=0x7fffe456ed58 sp=0x7fffe456eba8 pc=0x45a79d
runtime.morestack()
        /home/fornellas/.cache/resonance/GOROOT/go1.22.6.linux-amd64/src/runtime/asm_amd64.s:616 +0x7a fp=0x7fffe456ed60 sp=0x7fffe456ed58 pc=0x47235a

goroutine 1 gp=0xc0000061c0 m=0 mp=0x20b1960 [running]:
github.com/fornellas/resonance/internal/host.(*Docker).Type(0xc00015f2e0?)
        <autogenerated>:1 +0x3f fp=0xc0201e03a0 sp=0xc0201e0398 pc=0x821b5f
github.com/fornellas/resonance/internal/host.Docker.Type(...)
        /home/fornellas/src/resonance/internal/host/docker_unix.go:103
github.com/fornellas/resonance/internal/host.(*Docker).Type(0x0?)
        <autogenerated>:1 +0x29 fp=0xc0201e03b8 sp=0xc0201e03a0 pc=0x821b49
github.com/fornellas/resonance/internal/host.Docker.Type(...)
        /home/fornellas/src/resonance/internal/host/docker_unix.go:103
github.com/fornellas/resonance/internal/host.(*Docker).Type(0x0?)
        <autogenerated>:1 +0x29 fp=0xc0201e03d0 sp=0xc0201e03b8 pc=0x821b49
github.com/fornellas/resonance/internal/host.Docker.Type(...)
        /home/fornellas/src/resonance/internal/host/docker_unix.go:103
github.com/fornellas/resonance/internal/host.(*Docker).Type(0x0?)
        <autogenerated>:1 +0x29 fp=0xc0201e03e8 sp=0xc0201e03d0 pc=0x821b49
github.com/fornellas/resonance/internal/host.Docker.Type(...)
        /home/fornellas/src/resonance/internal/host/docker_unix.go:103
github.com/fornellas/resonance/internal/host.(*Docker).Type(0x0?)
        <autogenerated>:1 +0x29 fp=0xc0201e0400 sp=0xc0201e03e8 pc=0x821b49
github.com/fornellas/resonance/internal/host.Docker.Type(...)
        /home/fornellas/src/resonance/internal/host/docker_unix.go:103
github.com/fornellas/resonance/internal/host.(*Docker).Type(0x0?)
        <autogenerated>:1 +0x29 fp=0xc0201e0418 sp=0xc0201e0400 pc=0x821b49
...
```

Comment (fornellas) [reply cmd/lib.go:41]: PS: too tired to trust this, but I _think_ it should fix it:

```diff
diff --git a/internal/host/docker_unix.go b/internal/host/docker_unix.go
index e4716ab..c117876 100644
--- a/internal/host/docker_unix.go
+++ b/internal/host/docker_unix.go
@@ -100,7 +100,7 @@ func (d Docker) String() string {
 }
 
 func (d Docker) Type() string {
-       return d.Host.Type()
+       return "docker"
 }
 
 func (d Docker) Close() error {
diff --git a/internal/host/ssh.go b/internal/host/ssh.go
index 2999b48..fe222f1 100644
--- a/internal/host/ssh.go
+++ b/internal/host/ssh.go
@@ -427,7 +427,7 @@ func (s Ssh) String() string {
 }
 
 func (s Ssh) Type() string {
-       return s.Host.Type()
+       return "ssh"
 }
 
 func (s Ssh) Close() error {
```

TLDR: the host types that wrap something, must do the `Host.Type()` call, the "concrete types" (ssh, docker, localhost) which are "wrapable", must give the type directly.

**Review by fornellas — CHANGES_REQUESTED** [review]
@gfechio I think fixing that inifinite stack call there should get us to a merge \o/

**Comment by gfechio** [comment]
> @gfechio when you rebase, govulncheck job will fail, the fix is on #70.

#70 is merged now.

File: cmd/lib.go [hunk 22]
```diff
@@ -36,18 +33,31 @@ var defaultStoreHostTargetPath = "/var/lib/resonance"
...
+			}
+		}
+
+		if hst.Type() != "localhost" && !optionsMap["disable-agent"] {
+			hst, err = ihost.NewAgent(ctx, hst)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
```
Comment (fornellas) [inline cmd/lib.go:62]: @gfechio I'm really sorry to iterate this one more time, but I think you'll likely agree this is important. I hope I haven't missed anything here, if I did, **please** let me know (I'm sorry!). I recokn this is frustrating, and I promisse to only focus on important comments on future PRs: I'm sorry for what happened here.

If you check this comment: https://github.com/fornellas/resonance/pull/77/files#r1725719671, I suggested puling this outside of the `if options != ""` clause, because otherwise, when no option is passed, it'd _not_ be enabling the agent. We most definitely don't want that, as the agent can make things 25-90x faster (see benchmarks on #2). We touched the subject on a previous comment https://github.com/fornellas/resonance/pull/77#discussion_r1717481605 as well, where I explained the cases where the agent must be enabled.

WDYT?

```suggestion
	}
	if hst.Type() != "localhost" && !optionsMap["disable-agent"] {
		hst, err = ihost.NewAgent(ctx, hst)
		if err != nil {
			return nil, err
		}
	}
```

Comment (gfechio) [reply cmd/lib.go:62]: That makes sense, I think I might have missed it. :smile:

**Review by fornellas — APPROVED** [review]
